### ISSUE TITLE
BBC: Blead breaks MLEHMANN/Coro-6.57.tar.gz on debugging perls

### ISSUE THREAD

**Issue by andk** [Perl/perl5#20557]
**Description**

With v5.37.4-198-gdd66b1d793 tests started to fail for MLEHMANN/Coro-6.57.tar.gz on perls that have been built with -DEBUGGING=both. 

Sample fail report: http://www.cpantesters.org/cpan/report/46c3cf2e-6e27-11ed-9260-1e68f759ba8a

Failure diagnostics:
perl.c:2713: perl_run: Assertion `!PL_restartop' failed.

Bisect leads me to:

``` text
dd66b1d793c73fea9309c1d12a879369bf55bb83 is the first bad commit
commit dd66b1d793c73fea9309c1d12a879369bf55bb83
Author: Yves Orton <demerphq@gmail.com>
Date:   Wed Oct 12 18:50:25 2022 +0200

    pp_ctl.c - in try_yyparse do not leak PL_restartop from compile that dies

```

**Steps to Reproduce**

cpan -i MLEHMANN/Coro-6.57.tar.gz

**Expected behavior**

Should build, test and install the pkg

**Perl configuration**

```
# perl -V output goes here
Summary of my perl5 (revision 5 version 37 subversion 6) configuration:
  Commit id: d569e7cc1171d70b4ea183bc7c257dcf50f0ab12
  Platform:
    osname=linux
    osvers=5.19.0-2-amd64
    archname=x86_64-linux-thread-multi
    uname='linux k93msid 5.19.0-2-amd64 #1 smp preempt_dynamic debian 5.19.11-1 (2022-09-24) x86_64 gnulinux '
    config_args='-Dprefix=/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6/29fb -Dmyhostname=k93msid -Dinstallusrbinperl=n -Uversiononly -Dusedevel -des -Ui_db -Dlibswanted=cl pthread socket inet nsl gdbm dbm malloc dl ld sun m crypt sec util c cposix posix ucb BSD gdbm_compat -Duseithreads -Uuselongdouble -DEBUGGING=both'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2 -g'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='12.2.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib /usr/lib64
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/x86_64-linux-gnu/libc.so.6
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.35'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    DEBUGGING
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_TRACK_MEMPOOL
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Built under linux
  Compiled at Nov 20 2022 13:54:18
  %ENV:
    PERL5LIB=""
    PERL5OPT=""
    PERL5_CPANPLUS_IS_RUNNING="2346572"
    PERL5_CPAN_IS_RUNNING="2346572"
    PERL_CANARY_STABILITY_NOPROMPT="1"
    PERL_MM_USE_DEFAULT="1"
    PERL_USE_UNSAFE_INC="1"
  @INC:
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6/29fb/lib/site_perl/5.37.6/x86_64-linux-thread-multi
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6/29fb/lib/site_perl/5.37.6
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6/29fb/lib/5.37.6/x86_64-linux-thread-multi
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6/29fb/lib/5.37.6
    .
```

**Comment by demerphq** [comment]
On Tue, 29 Nov 2022 at 10:04, andk ***@***.***> wrote:

> *Description*
>
> With v5.37.4-198-gdd66b1d793 tests started to fail for
> MLEHMANN/Coro-6.57.tar.gz on perls that have been built with
> -DEBUGGING=both.
>
> Sample fail report:
> http://www.cpantesters.org/cpan/report/46c3cf2e-6e27-11ed-9260-1e68f759ba8a
>
> Failure diagnostics:
> perl.c:2713: perl_run: Assertion `!PL_restartop' failed.
>
> Bisect leads me to:
>
> dd66b1d793c73fea9309c1d12a879369bf55bb83 is the first bad commit
> commit dd66b1d793c73fea9309c1d12a879369bf55bb83
> Author: Yves Orton ***@***.***>
> Date:   Wed Oct 12 18:50:25 2022 +0200
>
>     pp_ctl.c - in try_yyparse do not leak PL_restartop from compile that dies
>
>
Just for some context:

    /* perl_parse() may end up starting its own run loops, which might end
     * up "leaking" PL_restartop from the parse phase into the run phase
     * which then ends up confusing run_body(). This leakage shouldn't
     * happen and if it does its a bug. */
    assert(!PL_restartop);

The general point of this assert is that any run loops started during
perl_parse() should complete and PL_restartop should be NULL by the time
run_body() starts the resulting program. PL_restartop for the record
represents the opcode that should execute if code died.

I will investigate. Maybe not too much today, I am a bit under the weather.

cheers,
Yves



-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by demerphq** [comment]
Thanks for the report. This should be fixed with #20558.

BTW, the code in Coro says this:

```
Coro-6.57$ grep -r -C1 PL_restartop
Coro/State.xs:    /* somebody or something will hit me for both perl_run and PL_restartop */
Coro/State.xs:    PL_restartop = PL_op;
Coro/State.xs-    perl_run (PL_curinterp);
```

maybe, I will leave that judgement to others. Marc isn't doing what that assert was intended to stop, and if it works for him then for now I see no reason to stand in his way, especially as moving the assert more perfectly captures my intent than the original did.

**Comment by demerphq** [comment]
Thanks again @andk for the report. Should be fixed now.
